File: include/apl/types/functions/builtins/fns/DShoeBuiltin.java
package APL.types.functions.builtins.fns;

import APL.types.*;
import APL.types.functions.Builtin;

import java.util.*;

@SuppressWarnings("Convert2Diamond") // convert.py chokes if not
public class DShoeBuiltin extends Builtin {
  @Override public std::string repr() {
    return "∪";
  }



  public Value call(Value w) {
    auto res = new LinkedHashSet<Value>(Arrays.asList(w.values()));
    return Arr.create(res.toArray(new Value[0]));
  }

  public Value call(Value a, Value w) {
    auto m = new LinkedHashSet<Value>(Arrays.asList(a.values()));
    m.addAll(Arrays.asList(w.values()));
    return Arr.create(m.toArray(new Value[0]));
  }
}

File: APL/types/functions/builtins/fns/DShoeBuiltin.cpp
#include <APL/DShoeBuiltin.h>
#include <APL/types/Arr.h>

namespace APL::types::functions::builtins::fns
{
	// using namespace APL::types;
	using Builtin = APL::types::functions::Builtin;

	std::string DShoeBuiltin::repr()
	{
	return L"∪";
	}

	std::shared_ptr<APL::types::Value> DShoeBuiltin::call(std::shared_ptr<APL::types::Value> w)
	{
	  auto res = std::make_shared<LinkedHashSet<std::shared_ptr<APL::types::Value>>>(Arrays::asList(w->values()));
	  return Arr::create(res->toArray(std::vector<std::shared_ptr<APL::types::Value>>(0)));
	}

	std::shared_ptr<APL::types::Value> DShoeBuiltin::call(std::shared_ptr<APL::types::Value> a, std::shared_ptr<APL::types::Value> w)
	{
	  auto m = std::make_shared<LinkedHashSet<std::shared_ptr<APL::types::Value>>>(Arrays::asList(a->values()));
	  m->addAll(Arrays::asList(w->values()));
	  return Arr::create(m->toArray(std::vector<std::shared_ptr<APL::types::Value>>(0)));
	}
}
